### PR TITLE
Add stream conformance tests for TranscodingStream

### LINKED ISSUE #44247: Encoder.Convert throws for charCount == 0
It seems like charCount == 0 should be a nop.  Instead it throws an exception about overflowing a buffer.

Same applies to Decoder.Convert.

I hit this while writing tests for TranscodingStream.

### PR DESCRIPTION
**Pull request by stephentoub**
And fix 0-length input handling of Encoder/Decoder.Convert (fixes https://github.com/dotnet/runtime/issues/44247)

cc: @GrabYourPitchforks, @tarekgh

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs [hunk 1]
```diff
@@ -261,6 +261,13 @@ public virtual unsafe int GetChars(ReadOnlySpan<byte> bytes, Span<char> chars, b
 
             bytesUsed = byteCount;
 
+            if (byteCount == 0)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: Do we need to handle `flush: true` here? The encoder may have a pending state that needs to be flushed even when `byteCount` is zero.

File: src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs [hunk 2]
```diff
@@ -261,6 +261,13 @@ public virtual unsafe int GetChars(ReadOnlySpan<byte> bytes, Span<char> chars, b
...
+            {
+                charsUsed = 0;
+                completed = _fallbackBuffer == null || _fallbackBuffer.Remaining == 0;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:267]: Write the fallback chars out when `charCount` is non-zero?

File: src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: I was thinking that'd be covered by just reporting completed as false, but if that's not the semantic, sure, I can change it.

Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: >             [](start = 0, length = 12)

There will be some concern I guess with this change which can cause infinite loops in the caller side if they didn't check before on the byte or char count and dependent on the exception especially when having flush = false.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: Is that a real concern and you think this should be closed?  Or just a hypothetical?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: With encoding I saw all weird things and I recall we ran into some infinite loop before in the full framework. So, I cannot be certain either way. if we want to go with this change, I would at least we document the change and use 6.0 release hopefully catch any bad usages of the API.

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs:264]: The behavior with _byteCount = 0, flush = true_ should be that the decoder attempts to flush its internal state to the destination, and if the destination is not large enough to hold the entire result then the operation should fail with an exception.

This matches the philosophy of the Decoder / Encoder types: they always try to make some forward progress. If they cannot make any forward progress at all (due to the destination buffer being too small), they fail.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/Decoder.cs]: I removed this commit and replaced it with one at the TranscodingStream level.  We can separately decide if/how to fix the issue in Encoder/Decoder.Convert.

**Review by tarekgh — APPROVED** [review]

**Pull request merged**